tools/cli/src/cli/lib.rs

```diff
@@ -139,3 +179,298 @@ pub(crate) fn password_prompt(prompt: &str) -> Option<String> {
     }
     None
 }
+
+pub const IDENTITY_UNAVAILABLE_ERROR_MESSAGE: &str = "The identity verification feature is not enabled for your account, please contact an administrator.";
+pub const CODE_FAILURE_ERROR_MESSAGE: &str = "The provided code doesn't match, please try again.";
+pub const INVALID_USER_ID_ERROR_MESSAGE: &str =
+    "account exists but cannot access the identity verification feature 😕";
+pub const INVALID_STATE_ERROR_MESSAGE: &str =
+    "The user identification flow is in an invalid state 😵😵";
+
+#[cfg(not(feature = "idv-tui"))]
+mod identify_user_no_tui {
+    use crate::{
+        CODE_FAILURE_ERROR_MESSAGE, IDENTITY_UNAVAILABLE_ERROR_MESSAGE,
+        INVALID_STATE_ERROR_MESSAGE, INVALID_USER_ID_ERROR_MESSAGE,
+    };
+    use async_recursion::async_recursion;
+    use kanidm_client::{ClientError, KanidmClient};
+    use kanidm_proto::internal::{IdentifyUserRequest, IdentifyUserResponse};
+
+    use dialoguer::{Confirm, Input};
+    use regex::Regex;
+    use std::{
+        io::{stdout, Write},
+        time::SystemTime,
+    };
+
+    lazy_static::lazy_static! {
+        pub static ref VALIDATE_TOTP_RE: Regex = {
+            #[allow(clippy::expect_used)]
+            Regex::new(r"^\d{5,6}$").expect("Failed to parse VALIDATE_TOTP_RE") // TODO: add an error ID (internal error, restart)
+        };
+    }
+
+    pub(super) enum IdentifyUserState {
+        Start,
+        IdDisplayAndSubmit,
+        SubmitCode,
+        DisplayCodeFirst { self_totp: u32, step: u32 },
+        DisplayCodeSecond { self_totp: u32, step: u32 },
+    }
+
+    fn server_error(e: ClientError) {
+        eprintln!("Server error!"); // TODO: add an error ID (internal error, restart)
+        eprintln!("{:?}", e);
+        println!("Exiting...");
+    }
+
+    #[async_recursion] // dear lord have mercy for I have sinned writing this function
```

Comment:
Yes, you have. Have you tried just having a loop and updating your state each loop iteration?

Reply:
I considered it but I couldn't overcome my feelings for purely functional stuff. Using `mut` in a function argument was too much for my little heart to handle :crying_cat_face:. I'll begrudgingly fix it though

Reply:
Rust isn't functional. And if you treat it like it is, you'll hit stack depth and crash. Sometimes the simple way, is the right way.

Reply:
Yes btw that was actually happening at one point lol. I was doing the totp refetch in the dumbest of ways by recursively calling the function and since I had some very poor logic going on it would make batch of 100 calls to itself .. let's say it didn't last more than 20 seconds. Lesson learned

Reply:
fixed!